go/vt/vtgate/planner.go

```diff
@@ -52,40 +53,80 @@ func NewPlanner(ctx context.Context, serv topo.SrvTopoServer, cell string, cache
 	return plr
 }
 
-// LoadVSchema loads the VSchema from the topo. The function does
+// WatchVSchema watches the VSchema from the topo. The function does
 // not return an error. It instead logs warnings on failure.
-func (plr *Planner) LoadVSchema(ctx context.Context) {
-	formal := &vindexes.VSchemaFormal{
-		Keyspaces: make(map[string]vindexes.KeyspaceFormal),
-	}
+// We get the list of keyspaces to watch at the beginning (because it's easier)
+// and we don't watch that. So when adding a new keyspace, for now,
+// vtgate needs to be restarted (after having rebuilt the serving graph).
+// This could be fixed by adding a WatchSrvKeyspaceNames API to topo.Server,
+// and when it triggers we'd diff the values, and terminate the missing
+// SrvKeyspaces, and start watching the new SrvKeyspaces.
+func (plr *Planner) WatchVSchema(ctx context.Context) {
 	keyspaces, err := plr.serv.GetSrvKeyspaceNames(ctx, plr.cell)
 	if err != nil {
 		log.Warningf("Error loading vschema: could not read keyspaces: %v", err)
 		return
 	}
-	for _, keyspace := range keyspaces {
-		formal.Keyspaces[keyspace] = vindexes.KeyspaceFormal{}
-		kschema, err := plr.serv.GetVSchema(context.TODO(), keyspace)
-		if err != nil {
-			log.Warningf("Error loading vschema for keyspace: %s: %v", keyspace, err)
-			continue
-		}
+
+	// we will wait until we get the first value for each keyspace
+	// before returning
+	var wg sync.WaitGroup
+
+	// mu protects formal
+	var mu sync.Mutex
+	formal := &vindexes.VSchemaFormal{
+		Keyspaces: make(map[string]vindexes.KeyspaceFormal),
+	}
+	processKeyspace := func(keyspace, kschema string) {
+		// Note we use a closure here to allow for defer()
+		// to work properly, and also to let the caller
+		// do more things even if this fails.
+
+		// unpack the new VSchema, or skip it
 		var kformal vindexes.KeyspaceFormal
-		err = json.Unmarshal([]byte(kschema), &kformal)
-		if err != nil {
-			log.Warningf("Error unmarshalling vschema for keyspace: %s: %v", keyspace, err)
-			continue
+		if err := json.Unmarshal([]byte(kschema), &kformal); err != nil {
+			log.Warningf("Error unmarshalling vschema for keyspace %s: %v", keyspace, err)
+			return
 		}
+
+		// rebuild the new component
+		mu.Lock()
+		defer mu.Unlock()
 		formal.Keyspaces[keyspace] = kformal
+		vschema, err := vindexes.BuildVSchema(formal)
```

Comment:
The downside of not re-reading everything is that we won't handle keyspace deletes correctly. I'm guessing it's not a big deal since it's pretty rare. Will it work as expected if a new keyspace was added?